segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,407 @@
+                    "current size of segment (%d).", offset, fileSize));
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return bytesRead;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getFilesystemRoot(), streamSegmentName),
+                PosixFileAttributes.class);
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                !(attrs.permissions().contains(OWNER_WRITE)), false,
+                new ImmutableDate(attrs.creationTime().toMillis()));
+
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        return Files.exists(Paths.get(config.getFilesystemRoot(), streamSegmentName));
+    }
+
+    @SneakyThrows
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        log.info("Creating Segment {}", streamSegmentName);
+        Set<PosixFilePermission> perms = new HashSet<>();
+        // add permission as rw-r--r-- 644
+        perms.add(PosixFilePermission.OWNER_WRITE);
+        perms.add(PosixFilePermission.OWNER_READ);
+        perms.add(PosixFilePermission.GROUP_READ);
+        perms.add(PosixFilePermission.OTHERS_READ);
+        FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        Files.createDirectories(path.getParent());
+        Files.createFile(path, fileAttributes);
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return this.syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        if (handle.isReadOnly()) {
+            throw new IllegalArgumentException("Write called on a readonly handle of segment "
+                    + handle.getSegmentName());
+        }
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+        //Fix for Jenkins as Jenkins runs as super user privileges. Jenkins runs with root privileges which means
```

Comment:
Let's not mention jenkins explicitly and say that when running with privileged mode, a write might succeed despite the permissions. We could jenkins as an example rather than say that this is for jenkins.

Another question is if this is strictly necessary. Trying to delete a file with sudo from the command line will ask for a confirmation to override if the write permission isn't set for the user. You sure this is necessary? I'm mostly curious as I think it is harmless to have this check.